src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs

```diff
@@ -300,8 +307,16 @@ private ReadOnlyMemory<byte> ReadChunkFromConnectionBuffer(int maxBytesToRead, C
                                     _state = ParsingState.Done;
                                     _connection.CompleteResponse();
                                     _connection = null;
+
                                     break;
                                 }
+                                // Parse the trailers.
+                                else if (!_connection.IsDisposed)
```

Comment:
I'm trying to understand how the connection can ever be disposed here.

As I understand it, we don't dispose the connection until after we attempt to drain, and only if the drain is unsuccessful (otherwise the connection would have been returned to the pool). 

It does seem like we probably need some sort of check here, because there's no need to actually process trailing headers if we are draining (though as far as I can see, there's no actual harm in it). But I don't think this check does it. Maybe we should check if the stream itself is disposed?

Reply:
This is a great find - thanks!

Here is a little bit more info: when we dispose the current response stream,
1. `_disposed` is set to true.
2. Drain the stream, which will call the current method (`ReadChunkFromConnectionBuffer`) to read to the end.
3. The final state in `ReadChunkFromConnectionBuffer` will always be `ConsumeTrailers`.

This check will avoid calling `ParseHeaderNameValue` in the above scenario.

Will change.